java/arcs/core/storage/StorageProxy.kt

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtException
+import arcs.core.crdt.CrdtModel
+import arcs.core.crdt.CrdtOperation
+import arcs.core.crdt.internal.VersionMap
+import arcs.core.util.guardWith
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/** ValueAndVersion is a tuple of a value of some type and a [VersionMap] provided from its [StorageProxy]. */
+data class ValueAndVersion<T>(val value: T, val versionMap: VersionMap)
+
+/**
+ * StorageProxy is an intermediary between a [Handle] and [Store].
+ * It provides up-to-date CRDT state to readers, and ensures write operations apply cleanly before
+ * forwarding to the store.
+ *
+ * @param T the consumer data type for the model behind this proxy
+ * @property model a concrete instance of the [CrdtModel] the proxy will use to keep a local copy.
+ */
+class StorageProxy<Data : CrdtData, Op : CrdtOperation, T>(
+    storeEndpointProvider: StorageCommunicationEndpointProvider<Data, Op, T>,
+    initialCrdt: CrdtModel<Data, Op, T>
+) {
+    private val mutex = Mutex()
+    private val handles: MutableSet<Handle<Data, Op, T>> by guardWith(mutex, mutableSetOf());
+    private val crdt by guardWith(mutex, initialCrdt)
+    private val waitingSyncs: MutableList<CompletableDeferred<ValueAndVersion<T>>> by guardWith(mutex, mutableListOf())
+    private var synchronized by guardWith(mutex, false)
+    private var listenerAttached by guardWith(mutex, false)
+    private var hasReaders by guardWith(mutex, false)
+
+    private val store = storeEndpointProvider.getStorageEndpoint()
+
+
+    /**
+     * Connects a handle. If the handle is readable, it will receive the configured callbacks
+     * specified by [HandleOptions].
+     */
+    suspend fun registerHandle(handle: Handle<Data, Op, T>): VersionMap {
+        // non-readers don't get callbacks, return early
+        if (!handle.canRead) {
+            mutex.withLock {
+                return crdt.versionMap.copy()
+            }
+        }
+
+        mutex.withLock() {
```

Comment:
It seems like the only reason we need to hold a lock here is to update the `handles` set and get the current version.  I think most of the chunks within can be brought outside the lock to remove the risk of deadlock.  Additionally, something like `requestSynchronization` might be better done as something which is self-contained and returns a deferred with whatever you need from it - rather than having it be a method with an opaque side-effect.